src/data/finset/functor.lean

```diff
@@ -0,0 +1,239 @@
+  comp_map := λ α β γ f g s, image_image.symm }
+
+@[simp] lemma fmap_def {s : finset α} (f : α → β) : f <$> s = s.image f := rfl
+
+end functor
+
+/-! ### Pure -/
+
+instance : has_pure finset := ⟨λ α x, {x}⟩
+
+@[simp] lemma pure_def {α} : (pure : α → finset α) = singleton := rfl
+
+/-! ### Applicative functor -/
+
+section applicative
+variables {α β : Type u} [Π P, decidable P]
+
+instance : applicative finset :=
+{ seq := λ α β t s, t.sup (λ f, s.image f),
+  seq_left := λ α β s t, if t = ∅ then ∅ else s,
+  seq_right := λ α β s t, if s = ∅ then ∅ else t,
+  .. finset.functor,
+  .. finset.has_pure }
+
+@[simp] lemma seq_def (s : finset α) (t : finset (α → β)) : t <*> s = t.sup (λ f, s.image f) := rfl
+@[simp] lemma seq_left_def (s : finset α) (t : finset β)  : s <* t = if t = ∅ then ∅ else s := rfl
+@[simp] lemma seq_right_def (s : finset α) (t : finset β)  : s *> t = if s = ∅ then ∅ else t := rfl
+
+instance : is_lawful_applicative finset :=
+{ seq_left_eq := λ α β s t, begin
+    rw [seq_def, fmap_def, seq_left_def],
+    obtain rfl | ht := t.eq_empty_or_nonempty,
+    { simp_rw [if_pos rfl, image_empty], exact (sup_bot _).symm },
+    { ext a,
+      rw [if_neg ht.ne_empty, mem_sup],
+      refine ⟨λ ha, ⟨const β a, mem_image_of_mem _ ha, mem_image_const_self.2 ht⟩, _⟩,
+      rintro ⟨f, hf, ha⟩,
+      rw mem_image at hf ha,
+      obtain ⟨b, hb, rfl⟩ := hf,
+      obtain ⟨_, _, rfl⟩ := ha,
+      exact hb }
+  end,
+  seq_right_eq := λ α β s t, begin
+    rw [seq_def, fmap_def, seq_right_def],
+    obtain rfl | hs := s.eq_empty_or_nonempty,
+    { rw [if_pos rfl, image_empty, sup_empty, bot_eq_empty] },
+    { ext a,
+      rw [if_neg hs.ne_empty, mem_sup],
+      refine ⟨λ ha, ⟨id, mem_image_const_self.2 hs, by rwa image_id⟩, _⟩,
+      rintro ⟨f, hf, ha⟩,
+      rw mem_image at hf ha,
+      obtain ⟨b, hb, rfl⟩ := ha,
+      obtain ⟨_, _, rfl⟩ := hf,
+      exact hb }
+  end,
+  pure_seq_eq_map := λ α β f s, sup_singleton,
+  map_pure := λ α β f a, image_singleton _ _,
+  seq_pure := λ α β s a, sup_singleton'' _ _,
+  seq_assoc := λ α β γ s t u, begin
+    ext a,
+    simp_rw [seq_def, fmap_def],
+    simp only [exists_prop, mem_sup, mem_image],
+    split,
+    { rintro ⟨g, hg, b, ⟨f, hf, a, ha, rfl⟩, rfl⟩,
+      exact ⟨g ∘ f, ⟨comp g, ⟨g, hg, rfl⟩, f, hf, rfl⟩, a, ha, rfl⟩ },
+    { rintro ⟨c, ⟨_, ⟨g, hg, rfl⟩, f, hf, rfl⟩, a, ha, rfl⟩,
+      exact ⟨g, hg, f a, ⟨f, hf, a, ha, rfl⟩, rfl⟩ }
+  end,
+  .. finset.is_lawful_functor }
+
+instance : is_comm_applicative finset :=
+{ commutative_prod := λ α β s t, begin
+    simp_rw [seq_def, fmap_def, sup_image, sup_eq_bUnion],
+    change s.bUnion (λ a, t.image $ λ b, (a, b)) = t.bUnion (λ b, s.image $ λ a, (a, b)),
+    transitivity s.product t;
+      [rw product_eq_bUnion, rw product_eq_bUnion_right]; congr; ext; simp_rw mem_image,
+  end,
+  .. finset.is_lawful_applicative }
+
+end applicative
+
+/-! ### Monad -/
+
+section monad
+variables [Π P, decidable P]
```

Comment:
What's the point in this vs just using classical?

Reply:
Being able to override it with a `decidable_eq` instance when we have it. I'm not so sure this is all that useful. I copied it from the design of `data.multiset.functor`.